src/neural/network_cudnn.cu

```diff
@@ -154,64 +153,63 @@ class ConvLayer : public BaseLayer {
   cudnnActivationDescriptor_t activation_;
 };
 
-class SoftMaxLayer : public BaseLayer {
+template <typename DataType>
+class SoftMaxLayer : public BaseLayer<DataType> {
  public:
   SoftMaxLayer(BaseLayer *ip);
-  void Eval(int N, float *output, const float *input, const float *input2,
-            float *scratch, cudnnHandle_t cudnn,
+  void Eval(int N, DataType *output, const DataType *input, const DataType *input2,
+            void *scratch, cudnnHandle_t cudnn,
             cublasHandle_t cublas) override;
 
  private:
   cudnnTensorDescriptor_t out_tensor_desc_;
 };
 
-class BNLayer : public BaseLayer {
+template <typename DataType>
+class BNLayer : public BaseLayer<DataType> {
  public:
   BNLayer(BaseLayer *ip, bool relu);
   ~BNLayer();
 
   void LoadWeights(float *cpuMeans, float *cpuVar);
-  void Eval(int N, float *output, const float *input, const float *input2,
-            float *scratch, cudnnHandle_t cudnn,
+  void Eval(int N, DataType *output, const DataType *input, const DataType *input2,
+            void *scratch, cudnnHandle_t cudnn,
             cublasHandle_t cublas) override;
 
  private:
   const bool use_relu_;
+
+  // always in float irrespective of DataType
```

Comment:
Sorry, for nitpicking, but consider changing all comments to proper English sentences. (starting with capital letter, ending with a period).

Also consider running `clang-format -I -style=google network_cudnn.cu`  but I can do that myself if you don't have clang installed.

Reply:
Don't have clang.